### ISSUE TITLE
fit_classifier_generator() should take a validation generator as an argument

### ISSUE THREAD

**Issue by zhao1701** [zhao1701/supervised-vae#1]
So to sum up what we wanted the fit_classifier_generator to do in terms of validating data:
- The validation_generator parameter should default to None
- If an actual generator is passed to validation_generator, then at the end of each epoch, we want three things printed out:
  - training accuracy
  - validation accuracy
  - validation loss
- We don't need to print out training loss, because that data is already being sent to Tensorboard

**Comment by serenazzz** [comment]
Added _calc_overall_metrics it calculates both accuracy and loss for a specific data generator. One note is that I printed training loss anyway since it comes with the function - but feel free to remove it in fit_classifier_generator()

**Comment by zhao1701** [comment]
Hey @serenazzz ,

Sorry it took so long for me to get around to looking at this. Thanks so much for adding the accuracy calculations as a new function rather than just making the fit functions more complicated. That's going to be super helpful. I like too that you're printing the results following each epoch. It means if we don't have Tensorboard, we're at least still seeing something.

I've got a few notes about cosmetics - just little stuff dealing with the PEP8 standard which would be fairly simple and a comment about something more structural.

### PEP8 notes
- **test_generator_functions.py**
  - Line 48: Shouldn't there be a space separating **val_gen** and the comma that comes before it?
- **__init__.py**
  - Can we rename **batch_accuracy_op** to **batch_accuracy_update**? I think that more explicitly explains what the operation actually does.
  - Line 5: Do we need the accuracy_score function from scikit-learn? I don't see it being used anywhere else in the code.
  - Line 335, 382: Docstrings should begin with a capital letter and end with periods (even when they're not complete sentences)
  - Line 340, 344, 345, 386, 387, 388: There should be a space before the colon. Also following the colon should be the data type, with a more descriptive explanation indented underneath.
  - Line 347: When passing arguments to functions, there should be no spaces on either side of the equal sign.
  - Line 347, 349: Can we choose a better variable scope than *my_metric*? How about just *accuracy*?
  - Line 361, 365: Single line comments should have a space following the *#* and be capitalized
  - Line 375: One space after the comma

Sorry if it seems nit-picky, but it's sort of like grammar. Imagine if someone else were to read our code. We want to feel good about the fact that it's consistent and not messy. If ever in doubt about how to space or format something, try taking a look at other parts of the code to see what the existing patterns are.

### Structural notes (all for __init__.py)
- Line 363, 367: Since we're feeding the same dictionary to both session run calls, maybe we should define the dictionary ahead of time so it doesn't have to be defined twice, one for both calls.
- I'm not sure that **batch_accuracy**, **batch_update_op**,  **running_vars** and **running_vars_initializer** need to be re-defined on every single call to **_calc_overall_metrics()**. They're just tensors whose values will change depending on what we pass to feed_dict. As such, I think they should be defined outside of the function and as object attributes so that the function still has access to them nonetheless. Perhaps consider writing a **_create_metrics()** function that defines these tensors ahead of time (we already have **_create_loss()**, **_create_optimizer()**, etc...)

Anyways, it looks like a lot, but most of it is just little formatting things. This shouldn't take very long, right? If anything's not clear, feel free to give me a ring. Thanks so much!

**Comment by zhao1701** [comment]
I think there might be one more issue in the way that we're calculating accuracy. **self.y_pred_denoised**, which we're feeding to **tf.metrics.accuracy** is a probability distribution and not the actual prediction. This means that the accuracy will basically be 0 always. What this means is that we should rename any current **predict** functions as  **predict_proba**, and write a new **predict** function that actual returns predicted labels. This also means we need a new tensor, something like **self.y_pred_labels** that is the result of some operation applied to **self.y_pred_denoised**.
